File: 백준/Gold/2636. 치즈/치즈.py
# 2636
# 치즈

import sys
from collections import deque

input = sys.stdin.readline

def bfs(start_x, start_y, cheeses):
    dx = [-1, 1, 0, 0] # 방향 벡터
    dy = [0, 0, -1, 1]

    queue = deque([(start_x, start_y)])
    visited = set([(start_x, start_y)])
    melt = deque([])

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < len(cheeses) and 0 <= ny < len(cheeses[0]):
                if (nx, ny) not in visited:
                    visited.add((nx, ny))
                    if cheeses[nx][ny] == 0: # 공기라면 계속 탐색해야하므로 queue에 추가
                        queue.append((nx, ny))
                    elif cheeses[nx][ny] == 1: # 치즈라면 녹여야하니까 melt에 추가
                        melt.append((nx, ny))

    
    for x, y in melt:
        cheeses[x][y] = 0
    return len(melt)

N, M = map(int, input().split(" "))

cheeses = []
cnt = 0

for i in range(N):
    cheeses.append(list(map(int, input().split())))
    cnt += sum(cheeses[i]) # 치즈 개수만 세기

time = 1

while True:
    meltCnt = bfs(0, 0, cheeses)
    cnt -= meltCnt
    if cnt == 0:
        print(time)
        print(meltCnt)
        break
    time += 1

File: 백준/Gold/2636. 치즈/치즈.java
import java.util.*;
import java.io.*;

public class Main {
	static int N, M;
	static int[][] cheeses;
	static int[] dx = {-1, 1, 0, 0};
	static int[] dy = {0, 0, 1, -1};
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		cheeses = new int[N][M];
		int cheesesCnt = 0;
		for(int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j = 0; j < M; j++) {
				cheeses[i][j] = Integer.parseInt(st.nextToken());
				if (cheeses[i][j] == 1) cheesesCnt++;
			}
		}
		
		int t = 0;
		int last = 0;
		while(true) {
			if (cheesesCnt == 0) break;
			last = bfs();
			cheesesCnt-=last;
			t++;
		}
		System.out.println(t);
		System.out.println(last);
	}
	
	public static int bfs() {
		boolean[][] visited = new boolean[N][M];
		Queue<int[]> q = new ArrayDeque<>();
		q.offer(new int[]{0, 0});
		visited[0][0] = true;
		List<int[]> melt = new ArrayList<>();
		
		while(!q.isEmpty()) {
			int[] now = q.poll();
			
			for(int i = 0; i < 4; i++) {
				int nx = now[0]+dx[i];
				int ny = now[1]+dy[i];
				
				if (!inRange(nx, ny) || visited[nx][ny]) continue;
				int[] tmp = new int[]{nx, ny};
				visited[nx][ny] = true;
				// 치즈면 지울거임
				if (cheeses[nx][ny] == 1) {
					melt.add(tmp);
				} else { // 아니면 이동할거임
					q.offer(tmp);
				}
			}
		}
		
		for(int[] p : melt) {
			cheeses[p[0]][p[1]] = 0;
		}
		
		return melt.size();
	}
	
	public static boolean inRange(int nx, int ny) {
		if (nx < 0 || nx >= N || ny < 0 || ny >= M) return false;
		return true;
	}

}
